amcl/src/amcl/amcl.cpp

```diff
@@ -0,0 +1,826 @@
+  if(frame_to_laser_.find(laser_scan->header.frame_id) == frame_to_laser_.end())
+  {
+    ROS_DEBUG("Setting up laser %d (frame_id=%s)\n", (int)frame_to_laser_.size(), laser_scan->header.frame_id.c_str());
+    lasers_.push_back(new AMCLLaser(*laser_));
+    lasers_update_.push_back(true);
+    laser_index = frame_to_laser_.size();
+
+    tf::Stamped<tf::Pose> ident (tf::Transform(tf::createIdentityQuaternion(),
+                                             tf::Vector3(0,0,0)),
+                                 ros::Time(), laser_scan->header.frame_id);
+    geometry_msgs::PoseStamped ident_pose;
+    tf::poseStampedTFToMsg(ident, ident_pose);
+    tf::Stamped<tf::Pose> laser_pose;
+    try
+    {
+      geometry_msgs::PoseStamped pose_out;
+      tf_buffer_.transform(ident_pose, pose_out, std::string(amclParams_.base_frame_id_));
+      tf::poseStampedMsgToTF(pose_out, laser_pose);
+    }
+    catch(tf::TransformException& e)
+    {
+      ROS_ERROR("Couldn't transform from %s to %s, "
+                "even though the message notifier is in use",
+                laser_scan->header.frame_id.c_str(),
+                amclParams_.base_frame_id_.c_str());
+      return;
+    }
+
+    pf_vector_t laser_pose_v;
+    laser_pose_v.v[0] = laser_pose.getOrigin().x();
+    laser_pose_v.v[1] = laser_pose.getOrigin().y();
+    // laser mounting angle gets computed later -> set to 0 here!
+    laser_pose_v.v[2] = 0;
+    lasers_[laser_index]->SetLaserPose(laser_pose_v);
+    ROS_DEBUG("Received laser's pose wrt robot: %.3f %.3f %.3f",
+              laser_pose_v.v[0],
+              laser_pose_v.v[1],
+              laser_pose_v.v[2]);
+
+    frame_to_laser_[laser_scan->header.frame_id] = laser_index;
+  } else {
+    // we have the laser pose, retrieve laser index
+    laser_index = frame_to_laser_[laser_scan->header.frame_id];
+  }
+
+  // Where was the robot when this scan was taken?
+  pf_vector_t pose;
+  if(!getOdomPose(latest_odom_pose_, pose.v[0], pose.v[1], pose.v[2],
+                  laser_scan->header.stamp, amclParams_.base_frame_id_))
+  {
+    ROS_ERROR("Couldn't determine robot's pose associated with laser scan");
+    return;
+  }
+
+  pf_vector_t delta = pf_vector_zero();
+
+  if(pf_init_)
+  {
+    // Compute change in pose
+    //delta = pf_vector_coord_sub(pose, pf_odom_pose_);
+    delta.v[0] = pose.v[0] - pf_odom_pose_.v[0];
+    delta.v[1] = pose.v[1] - pf_odom_pose_.v[1];
+    delta.v[2] = angle_diff(pose.v[2], pf_odom_pose_.v[2]);
+
+    // See if we should update the filter
+    bool update = fabs(delta.v[0]) > amclParams_.d_thresh_ ||
+                  fabs(delta.v[1]) > amclParams_.d_thresh_ ||
+                  fabs(delta.v[2]) > amclParams_.a_thresh_;
+    update = update || force_update_;
+    force_update_ = false;
+
+    // Set the laser update flags
+    if(update)
+      for(unsigned int i=0; i < lasers_update_.size(); i++)
+        lasers_update_[i] = true;
+  }
+
+  bool force_publication = false;
+  if(!pf_init_)
```

Comment:
split into functions!

Reply:
change the logic of pf init so that laser callback doesn't initialize it secretly

Reply:
https://6river.atlassian.net/wiki/spaces/SEng/pages/2774335710/AMCL+API reason why this can't be changed